src/hotspot/share/utilities/byteswap.hpp

```diff
@@ -0,0 +1,176 @@
+/*
+ * Copyright (c) 2023, Google and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#ifndef SHARE_UTILITIES_BYTESWAP_HPP
+#define SHARE_UTILITIES_BYTESWAP_HPP
+
+// Byte swapping for 16-bit, 32-bit, and 64-bit integers.
+
+// byteswap<T>()
+//
+// Reverses the bytes for the value of the integer type T. Partially compatible with std::byteswap
+// introduced in C++23
+
+#include "metaprogramming/enableIf.hpp"
+#include "utilities/debug.hpp"
+#include "utilities/globalDefinitions.hpp"
+#include "utilities/macros.hpp"
+
+#if defined(__linux__)
+#include <byteswap.h>
+#endif
+
+#if defined(__FreeBSD__) || defined(__NetBSD__) || defined(__DragonFly__)
+#include <sys/endian.h>
+#endif
+
+#if defined(__OpenBSD__)
+#include <endian.h>
+#endif
+
+#if defined(__APPLE__)
+#include <libkern/OSByteOrder.h>
+#endif
+
+#include <cstddef>
+#include <cstdint>
+#include <cstdlib>
+#include <type_traits>
+
+#if defined(_MSC_VER)
```

Comment:
I'd rather see OS specific header files to do all this than pollute what is supposed to be shared code with all this OS specific stuff.

Reply:
> I'd rather see OS specific header files to do all this than pollute what is supposed to be shared code with all this OS specific stuff.

Definitely.

Reply:
Doubling back to this thread, I have now removed all the OS-specific stuff as I mentioned elsewhere. It was unnecessarily pedantic on my part to include it, considering all those platforms use GCC/Clang, MSVC, or XL C/C++.